### PR TITLE
Turn Clipboard into a List

### PR DESCRIPTION
**Pull request by freddyguythe10th**
Just turns a clipboard into a List. I had trouble getting the environment set up, but with my testing it seems to work pretty well. Also updated the REPL Help section to include updated information about the clipboard.

### CODE REVIEW

File: src/UI/Panels/ClipboardPanel.cs [hunk 1]
```diff
@@ -4,8 +4,9 @@ namespace UnityExplorer.UI.Panels
 {
     public class ClipboardPanel : UEPanel
     {
-        public static object Current { get; private set; }
-
+        public static List<object> ClipboardItems = new List<object>();
```
Comment (yukieiji) [inline src/UI/Panels/ClipboardPanel.cs:7]: I read the code and it is in FILO logic, so is it possible to refactor it using `System.Collection.Generic.Stack<object>`?
If it's possible to refactor it, it would eliminate the need for boundary checks, etc.

File: src/UI/Panels/ClipboardPanel.cs [hunk 2]
```diff
@@ -93,21 +151,60 @@ protected override void ConstructPanelContent()
...
-            UniverseLib.UI.Models.ButtonRef inspectButton = UIFactory.CreateButton(currentPasteHolder, "InspectButton", "Inspect");
+            clearButton.OnClick += ClearClipboard;
+            
+            // Inspect Selected button
+            UniverseLib.UI.Models.ButtonRef inspectButton = UIFactory.CreateButton(firstRow, "InspectButton", "Inspect");
             UIFactory.SetLayoutElement(inspectButton.Component.gameObject, minHeight: 25, flexibleHeight: 0, minWidth: 80, flexibleWidth: 0);
             inspectButton.OnClick += InspectClipboard;
+            
+            // Add New button
+            UniverseLib.UI.Models.ButtonRef addNewButton = UIFactory.CreateButton(firstRow, "AddNewButton", "Add");
```
Comment (yukieiji) [inline src/UI/Panels/ClipboardPanel.cs:162]: Is this a test code? If so, please remove it.

File: src/UI/Panels/ClipboardPanel.cs [hunk 3]
```diff
@@ -19,25 +20,76 @@ public class ClipboardPanel : UEPanel
...
 
+        public static void SelectClipboardItem(int index)
+        {
+            selectedItem = index;
+        }
+
+        /// <summary>
+        /// Adds a new entry and selects it
+        /// </summary>
+        public static void AddNewEntry()
```
Comment (yukieiji) [inline src/UI/Panels/ClipboardPanel.cs:37]: Is this a test function? If so, please remove it.

File: src/UI/Panels/ClipboardPanel.cs [hunk 4]
```diff
@@ -19,25 +20,76 @@ public class ClipboardPanel : UEPanel
...
-            Notification.ShowMessage("Copied!");
+            if (obj == null || ClipboardItems.Count <= selectedItem)
+            {
+                ExplorerCore.Log($"The selected item index is outside the range! {ClipboardItems.Count} is the count.");
+                return;
+            }
+            Copy(obj,selectedItem, true);
+        }
+
+        public static void Copy(object obj, int index, bool showNotification)
```
Comment (yukieiji) [inline src/UI/Panels/ClipboardPanel.cs:53]: Is the `showNotification` arg necessary?

File: src/CSConsole/ScriptInteraction.cs [hunk 5]
```diff
@@ -31,9 +31,18 @@ public static void Stop(Coroutine coro)
 
         public static void Copy(object obj)
             => ClipboardPanel.Copy(obj);
+        
+        public static void Copy(object obj, int index)
```
Comment (yukieiji) [inline src/CSConsole/ScriptInteraction.cs:35]: I think this function is not used and should be removed.

File: src/CSConsole/ScriptInteraction.cs [hunk 6]
```diff
@@ -31,9 +31,18 @@ public static void Stop(Coroutine coro)
...
+            => ClipboardPanel.Copy(obj,index,false);
+        
+        public static void CopyToEnd(object obj)
```
Comment (yukieiji) [inline src/CSConsole/ScriptInteraction.cs:38]: I think this function is not used and should be removed.

File: src/UI/Panels/ClipboardPanel.cs [hunk 7]
```diff
@@ -19,25 +20,76 @@ public class ClipboardPanel : UEPanel
...
+            {
+                ClipboardItems.Insert(index,obj);
+            }
+            ClipboardItems[index] = obj;
+            if(showNotification)
+                Notification.ShowMessage("Copied to clipboard at index " + selectedItem + "!");
+            UpdateCurrentPasteInfo();
+        }
+
+        public static void CopyToEnd(object obj)
```
Comment (yukieiji) [inline src/UI/Panels/ClipboardPanel.cs:69]: I think this function is not used and should be removed.

File: src/UI/Panels/ClipboardPanel.cs [hunk 8]
```diff
@@ -4,8 +4,9 @@ namespace UnityExplorer.UI.Panels
 {
     public class ClipboardPanel : UEPanel
     {
-        public static object Current { get; private set; }
-
+        public static List<object> ClipboardItems = new List<object>();
+        public static int selectedItem = 0;
+        public static ClipboardPanel Instance;
```
Comment (yukieiji) [inline src/UI/Panels/ClipboardPanel.cs:9]: I don't think it needs to be a singleton.

File: src/UnityExplorer.csproj [hunk 9]
```diff
@@ -354,4 +354,9 @@
       <Private>False</Private>
     </Reference>
   </ItemGroup>
+  <ItemGroup>
+    <Content Include="..\UniverseLib\src\UniverseLib.csproj">
+      <Link>UniverseLib.csproj</Link>
+    </Content>
+  </ItemGroup>
```
Comment (yukieiji) [inline src/UnityExplorer.csproj:361]: Is this change necessary?

File: src/UnityExplorer.sln [hunk 10]
```diff
@@ -48,6 +50,32 @@ Global
...
+		{E3A1056F-8C8C-4394-A836-4803D6208AEA}.Release_ML_Cpp_net6preview|Any CPU.ActiveCfg = Release_Mono|Any CPU
+		{E3A1056F-8C8C-4394-A836-4803D6208AEA}.Release_ML_Cpp_net6preview|Any CPU.Build.0 = Release_Mono|Any CPU
+		{E3A1056F-8C8C-4394-A836-4803D6208AEA}.Release_ML_Mono|Any CPU.ActiveCfg = Release_Mono|Any CPU
+		{E3A1056F-8C8C-4394-A836-4803D6208AEA}.Release_ML_Mono|Any CPU.Build.0 = Release_Mono|Any CPU
+		{E3A1056F-8C8C-4394-A836-4803D6208AEA}.Release_STANDALONE_Cpp_CoreCLR|Any CPU.ActiveCfg = Release_Mono|Any CPU
+		{E3A1056F-8C8C-4394-A836-4803D6208AEA}.Release_STANDALONE_Cpp_CoreCLR|Any CPU.Build.0 = Release_Mono|Any CPU
+		{E3A1056F-8C8C-4394-A836-4803D6208AEA}.Release_STANDALONE_Cpp|Any CPU.ActiveCfg = Release_Mono|Any CPU
+		{E3A1056F-8C8C-4394-A836-4803D6208AEA}.Release_STANDALONE_Cpp|Any CPU.Build.0 = Release_Mono|Any CPU
+		{E3A1056F-8C8C-4394-A836-4803D6208AEA}.Release_STANDALONE_Mono|Any CPU.ActiveCfg = Release_Mono|Any CPU
+		{E3A1056F-8C8C-4394-A836-4803D6208AEA}.Release_STANDALONE_Mono|Any CPU.Build.0 = Release_Mono|Any CPU
```
Comment (yukieiji) [inline src/UnityExplorer.sln:78]: Is this change necessary?

File: src/UI/Panels/ClipboardPanel.cs [hunk 11]
```diff
@@ -93,21 +151,60 @@ protected override void ConstructPanelContent()
...
+
+            RefreshList();
+        }
+
+        public void RefreshList()
+        {
+            foreach (GameObject go in clipboardRows)
+            {
+                //Remove the row
+                GameObject.Destroy(go);
```
Comment (yukieiji) [inline src/UI/Panels/ClipboardPanel.cs:174]: I don't think it's a good idea to recreate it every time, so can't you use `CreateVerticalGroup` to add elements, etc.?

**Comment by yukieiji** [comment]
Code reviews were done.
It might be better to prepare Linux sh files for builds for compatibility.

File: src/UnityExplorer.sln (hunk 10 above)
Comment (freddyguythe10th) [reply src/UnityExplorer.sln:78]: This was done automatically by my IDE, so I don't really know.

File: src/UnityExplorer.csproj (hunk 9 above)
Comment (freddyguythe10th) [reply src/UnityExplorer.csproj:361]: This was done automatically by my IDE so references weren't broken while programming. It can probably be removed and still be built fine.

File: src/UI/Panels/ClipboardPanel.cs (hunk 7 above)
Comment (freddyguythe10th) [reply src/UI/Panels/ClipboardPanel.cs:69]: I made the CopyToEnd() method for when using the console, just because it makes it easier to add a new item and fill it at the end of the clipboard list. It could probably be removed, I just think it's useful for quick usage of the console.

File: src/UI/Panels/ClipboardPanel.cs (hunk 4 above)
Comment (freddyguythe10th) [reply src/UI/Panels/ClipboardPanel.cs:53]: This is just so if you run Copy() from the console it doesn't show the notification. This is mostly useful if you are copying something every second in a coroutine or something. We do want to still use notifications though for when a user clicks the "Copy To Clipboard" button in the inspector.

File: src/UI/Panels/ClipboardPanel.cs (hunk 3 above)
Comment (freddyguythe10th) [reply src/UI/Panels/ClipboardPanel.cs:37]: This adds a new entry to the end of clipboard and selects it. It's mostly used by the UI, but is necessary if you want to stay away from using the Console for everything.

File: src/UI/Panels/ClipboardPanel.cs (hunk 2 above)
Comment (freddyguythe10th) [reply src/UI/Panels/ClipboardPanel.cs:162]: This adds the button in the clipboard that actually adds a new entry. I can send a screen shot of what the clipboard looks like now.

**Comment by freddyguythe10th** [comment]
image
This is what the new clipboard looks like. The ```UniverseLib.UI.Models.ButtonRef addNewButton = UIFactory.CreateButton(firstRow, "AddNewButton", "Add");``` is what places down the "Add" button in the clipboard, and is responsible for adding new entries while using the UI. The ```AddNewEntry()``` method just adds an item to the end of the clipboard and selects it, making it easier to use with UI. It might not be necessary, as we could probably use delegates, but this helps separate stuff and makes more sense to me.

Sorry if these changes aren't the most clear, this is my first time doing a pull request. Hopefully this comment helped clear up some confusion.

File: src/UI/Panels/ClipboardPanel.cs (hunk 11 above)
Comment (freddyguythe10th) [reply src/UI/Panels/ClipboardPanel.cs:174]: I don't think redrawing it is the most efficient, but I'm not sure how we could do checks to make sure the list UI is always up to date with the actual list. For example, if the GameObject at index 0 is destroyed, how do we make sure that gets updated in the UI? I thought it might be more efficient to just redraw the UI instead of having a List of UI objects and doing stuff with those.

If you don't think it would be more efficient, just lemme know and we can probably figure out a better solution.

File: src/UI/Panels/ClipboardPanel.cs (hunk 8 above)
Comment (freddyguythe10th) [reply src/UI/Panels/ClipboardPanel.cs:9]: The reason I have this be a singleton is so then I can redraw the UI from static methods. For example, the Copy() method is static, because we want to be able to call that from the C# Console, but the UI elements are not static. The instance is just so we can redraw the UI from a static method.
If this needs to be changed, just lemme know and I can figure something out.

File: src/CSConsole/ScriptInteraction.cs (hunk 6 above)
Comment (freddyguythe10th) [reply src/CSConsole/ScriptInteraction.cs:38]: I made the CopyToEnd() method for when using the console, just because it makes it easier to add a new item and fill it at the end of the clipboard list. It could probably be removed, I just think it's useful for quick usage of the console.

File: src/CSConsole/ScriptInteraction.cs (hunk 5 above)
Comment (freddyguythe10th) [reply src/CSConsole/ScriptInteraction.cs:35]: This method allows you to replace an object at any index with a different object. It probably isn't necessary because I introduced the ```Clipboard``` object in the Console, I just thought maybe it would be useful to have.
If this needs to change, just lemme know and we can figure something out.

File: src/UI/Panels/ClipboardPanel.cs (hunk 1 above)
Comment (freddyguythe10th) [reply src/UI/Panels/ClipboardPanel.cs:7]: I've never used a ```Stack``` before. I'll look into it.

**Comment by freddyguythe10th** [comment]
I can make any changes you feel like might be best, as it's probably best to make sure the Clipboard is still just as easy to use as the original.

**Comment by yukieiji** [comment]
Thank you for your reply
It might be better to have an `Inspect` button on each row

**Comment by yukieiji** [comment]
Is the following work to be done by the user?

1. Press `Add` button
2. Press `Select` button on the copy target row
3. Press  `Copy to Clipboard` button in Object Explorer, etc.

If that is the case, it might be better to automatically add the row when you press `Copy to Clipboard` and not have an `Add` button.

File: src/UnityExplorer.sln (hunk 10 above)
Comment (yukieiji) [reply src/UnityExplorer.sln:78]: Okay, please revert before the merge for consistency in build logic, etc.

File: src/UI/Panels/ClipboardPanel.cs (hunk 4 above)
Comment (yukieiji) [reply src/UI/Panels/ClipboardPanel.cs:53]: I think it would be good to have a notification if the copied objects, etc. are displayed when copied.
Are there any problems with sending out a large number of notifications in current?

**Comment by yukieiji** [comment]
My personal opinion of the implementation is as follows.

1. each time “Copy to Clipboard” etc. is pressed, an object is added to “Clipboard” (FILO, last one added appears at the top).
2. only “Clear” button exists in “Clipboard“
3. in the case of “Clipboard” with objects, “Remove” button exist for each object.
4. `Paste()` function paste the last added object, `Paste(int index)` paste objects at the index of “Clipboard“

This implementation similar to the behavior of Clipboard in Windows.

File: src/UI/Panels/ClipboardPanel.cs (hunk 11 above)
Comment (yukieiji) [reply src/UI/Panels/ClipboardPanel.cs:174]: I feel the implementation of `InspectorManager.CreateInspector` would be helpful
It seems to me that by following that implementation, the Singleton class would not need to be implemented (it could be avoided by making the list part(`CreateVerticalGroup` gameobject) that holds the items static).

**Comment by freddyguythe10th** [comment]
I'm gonna close this pull request, and I'll add it back later when I make those changes.
